File: problems/p02/valid_parentheses.py
def is_valid(s: str) -> bool:
    """
    [REQUIRED] Why stack? list edge cases; Big-O analysis.
    """
    pairs = {')':'(', ']':'[', '}':'{'}
    st = []
    for ch in s:
        if ch in "([{":
            st.append(ch)
        elif ch in pairs:
            if not st or st[-1] != pairs[ch]:
                return False
            st.pop()
        else:
            # any other char -> invalid per problem definition
            return False
    return len(st) == 0

File: problems/p02/valid_parentheses.js
// [REQUIRED] Why stack? list edge cases; Big-O analysis.
function isValid(s) {
  const pairs = {')':'(', ']':'[', '}':'{'};
  const st = [];
  for (const ch of s) {
    if (ch === '(' || ch === '[' || ch === '{') st.push(ch);
    else if (ch in pairs) {
      if (st.length === 0 || st[st.length - 1] !== pairs[ch]) return false;
      st.pop();
    } else return false;
  }
  return st.length === 0;
}
module.exports = { isValid };
